include/alpaka/block/shared/dyn/BlockSharedMemDynMember.hpp

```diff
@@ -0,0 +1,128 @@
+                    auto operator=(BlockSharedMemDynMember const &) -> BlockSharedMemDynMember & = delete;
+                    //-----------------------------------------------------------------------------
+                    auto operator=(BlockSharedMemDynMember &&) -> BlockSharedMemDynMember & = delete;
+                    //-----------------------------------------------------------------------------
+                    /*virtual*/ ~BlockSharedMemDynMember() = default;
+
+                    uint8_t* dynMemBegin() const {return m_mem.data();}
+
+                    /*! \return the pointer to the begin of data after the portion allocated as dynamical shared memory.
+                     *!
+                     *! \tparam TDataAlignBytes Round-up offset to multiple of
+                     *! this number of bytes to maintain alignment. This
+                     *! alignment assumes, that the compiler places the instance
+                     *! at architecture-appropriate boundaries.
+                     */
+                    template<unsigned int TDataAlignBytes = core::vectorization::defaultAlignment>
+                    uint8_t* staticMemBegin() const
+                    {
+                        return m_mem.data() +
+                            (m_dynSize/TDataAlignBytes + (m_dynSize%TDataAlignBytes>0)*TDataAlignBytes);
+                    }
+
+                    /*! \return the remaining capacity for static block shared memory.
+                     *!
+                     *! \tparam TDataAlignBytes Multiple of bytes the static offset was rounded-up to
+                     */
+                    template<unsigned int TDataAlignBytes = 4>
+                    unsigned int staticMemCapacity() const
+                    {
+                        return staticAllocBytes -
+                            (m_dynSize/TDataAlignBytes + (m_dynSize%TDataAlignBytes>0)*TDataAlignBytes);
+                    }
+
+                private:
+                    static constexpr unsigned int staticAllocBytes = TStaticAllocKiB<<10;
+
+                    mutable std::array<uint8_t, staticAllocBytes> m_mem;
+                    unsigned int m_dynSize;
+                };
+#if BOOST_COMP_MSVC || defined(BOOST_COMP_MSVC_EMULATED)
+    #pragma warning(pop)
+#endif
+
+                namespace traits
+                {
+                    //#############################################################################
+                    template<
+                        typename T,
+                        unsigned int TStaticAllocKiB>
+                    struct GetMem<
+                        T,
+                        BlockSharedMemDynMember<TStaticAllocKiB>>
+                    {
+#if BOOST_COMP_GNUC
+    #pragma GCC diagnostic push
+    #pragma GCC diagnostic ignored "-Wcast-align" // "cast from 'unsigned char*' to 'unsigned int*' increases required alignment of target type"
```

Comment:
This warning is not generated without reason. It is undefined if the pointer resulting from this cast can be accessed or not. There are systems out there where you can access variables only when they are aligned correctly for their size (e.g. 4 byte alignment for `int`). On other systems accessing those variables works but is much slower than aligned access.
We are aligning the class (and so the array) with `defaultAlignment` so we can only correctly access types with this alignment requirement correctly.
 In `BlockSharedMemDynBoostAlignedAlloc` we added a check to verify that it is used correctly:
```
static_assert(
    core::vectorization::defaultAlignment >= alignof(T),
    "Unable to get block shared dynamic memory for types with alignment higher than defaultAlignment!");
```
We should probably add this check here.

Reply:
On other architectures, like GPUs (with 32-bit wide banks in shared memory) we want to be able to align types to smaller boundaries. I doubt that  `alignof` could capture this even in theory as it cannot know wether we are storing a value in global or shared memory.

Reply:
Apparently I was wrong about the GPU: If I reinterpret_cast a double pointer to shared memory on a Kepler generation GPU which is not 8-byte aligned, reading from (or writing) to it results int wrong values.

In this case I think, we should solve this problem a bit differently: Each allocation should be aligned to the larger of `TDataAlignBytes` and `alignof(T)`.

We need the assert for Dyn and St and I need to change the alignment of the  St section.